packages/core/src/api/app/FeatureFlags.tsx

```diff
@@ -0,0 +1,177 @@
+    );
+  }
+}
+
+/**
+ * The UserFlags class.
+ *
+ * This acts as a data structure for the user's feature flags. You
+ * can use this to retrieve, add, edit, delete, clear and save the user's
+ * feature flags to the local browser for persisted storage.
+ */
+export class UserFlags extends Map<FeatureFlagName, FeatureFlagState> {
+  constructor() {
+    validateBrowserCompat();
+
+    try {
+      const jsonString = window.localStorage.getItem('featureFlags') as string;
+      const json = JSON.parse(jsonString);
+      super(Object.entries(json));
+    } catch (err) {
+      super([]);
+    }
+  }
+
+  get(name: FeatureFlagName): FeatureFlagState {
+    validateFlagName(name);
+    return super.get(name) || FeatureFlagState.NotEnabled;
+  }
+
+  set(name: FeatureFlagName, state: FeatureFlagState): this {
+    validateFlagName(name);
+    super.set(name, state);
+    this.save();
+    return this;
+  }
+
+  delete(name: FeatureFlagName): boolean {
+    const exists = !!this.get(name);
+    super.delete(name);
+    this.save();
+    return exists;
+  }
+
+  clear(): void {
+    super.clear();
+    this.save();
+  }
+
+  save(): void {
+    window.localStorage.setItem(
+      'featureFlags',
+      JSON.stringify(this.toObject()),
+    );
+  }
+
+  toObject() {
+    return Array.from(this.entries()).reduce(
+      (obj, [key, value]) => ({ ...obj, [key]: value }),
+      {},
+    );
+  }
+
+  toJSON() {
+    return JSON.stringify(this.toObject());
+  }
+
+  toString() {
+    return this.toJSON();
+  }
+}
+
+/**
+ * The FeatureFlagsRegistry class.
+ *
+ * This acts as a holding data structure for feature flags
+ * that plugins wish to register for use in Backstage.
+ */
+export interface FeatureFlagsRegistryItem {
+  pluginId: string;
+  name: FeatureFlagName;
+}
+
+export class FeatureFlagsRegistry extends Array<FeatureFlagsRegistryItem> {
+  static from(entries: FeatureFlagsRegistryItem[]) {
+    Array.from(entries).forEach(entry => validateFlagName(entry.name));
+    return new FeatureFlagsRegistry(...entries);
+  }
+
+  push(...entries: FeatureFlagsRegistryItem[]): number {
+    Array.from(entries).forEach(entry => validateFlagName(entry.name));
+    return super.push(...entries);
+  }
+
+  concat(
+    ...entries: (
+      | FeatureFlagsRegistryItem
+      | ConcatArray<FeatureFlagsRegistryItem>
+    )[]
+  ): FeatureFlagsRegistryItem[] {
+    const _concat = super.concat(...entries);
+    Array.from(_concat).forEach(entry => validateFlagName(entry.name));
+    return _concat;
+  }
+
+  toObject() {
+    return [...this.values()];
+  }
+
+  toJSON() {
+    return JSON.stringify(this.toObject());
+  }
+
+  toString() {
+    return this.toJSON();
+  }
+}
+
+/**
+ * Create the FeatureFlags implementation based on the API.
+ */
+export class FeatureFlags implements FeatureFlagsApi {
+  public registeredFeatureFlags: FeatureFlagsRegistryItem[] = [];
+
+  getFlags(): UserFlags {
+    return new UserFlags();
```

Comment:
Better store this in a readonly member instead of constructing a new one on each call.

Reply:
Fixed in 0a4ad0f24cc6736096f25e17c6fb6c3acbdfae86